datafusion/physical-expr/src/aggregate/average.rs

```diff
@@ -383,6 +435,189 @@ impl RowAccumulator for AvgRowAccumulator {
     }
 }
 
+/// An accumulator to compute the average of `[PrimitiveArray<T>]`.
+/// Stores values as native types, and does overflow checking
+///
+/// F: Function that calcuates the average value from a sum of
+/// T::Native and a total count
+#[derive(Debug)]
+struct AvgGroupsAccumulator<T, F>
+where
+    T: ArrowNumericType + Send,
+    F: Fn(T::Native, u64) -> Result<T::Native> + Send,
+{
+    /// The type of the internal sum
+    sum_data_type: DataType,
+
+    /// The type of the returned sum
+    return_data_type: DataType,
+
+    /// Count per group (use u64 to make UInt64Array)
+    counts: Vec<u64>,
+
+    /// Sums per group, stored as the native type
+    sums: Vec<T::Native>,
```

Comment:
Is it possible to combine the `counts` and `sums` into one property, like `avg_states: Vec<(T::Native, u64)>`

Reply:
@alamb sounds like a useful suggestion

Reply:
> Is it possible to combine the counts and sums into one property, like avg_states: Vec<(T::Native, u64)>? Since one sum and the related count are always used together, I think it's better to put them together for better cache locality.

Thank you for the comment @yahoNanJing 

The reason the sums and counts are stored separately is to minimize copying when forming the final output -- since the final output is columnar (two columns) keeping the data as two `Vec`s allows the final `ArrayRefs` to be created directly from that data. 

It would be an interesting experiment to see if keeping them together and improving cache locality outweighed the extra copy.

BTW if people are looking to optimize the inner loops more, I think removing the bounds checks with unsafe might also help (but I don't plan to pursue it until I find need to optimize more)

So instead of 
```rust
                let sum = &mut self.sums[group_index];
                *sum = sum.add_wrapping(new_value);
```

```rust
                unsafe {
                  let sum = sums.get_unchecked_mut(group_index);
                  *sum = sum.add_wrapping(new_value);
                }
```

Reply:
Is it possible to make a tuple `(T::Native, u64)` as a primitive type at the arrow-rs side? Then we don't need to return two arrays for the `state`

Reply:
Ah -- I see what you are saying -- I think we could potentially use a `StructArray` for the state (which would be a single "column" in arrow) but the underlying storage is still two separate contiguous arrays. 

Maybe we could use `FixedSizeBinaryArray` 🤔  and pack/unpack the tuples to the appropriate size

It would be an interesting experiment

Reply:
I'm afraid both of the `StructArray` and `FixedSizeBinaryArray` may have additional overhead. 

If `T::Native` can be a tuple, then we can provide a new array, called `TupleArray`. The element type is a tuple, (T::Native, T::Native). Then this tuple can be any nested tuples. And this new `TupleArray` can cover any nested tuple cases.

Reply:
It would definitely be a cool thing to try